test/temporary_directory_fixture.hpp

```diff
@@ -0,0 +1,100 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TEMPORARY_DIRECTORY_FIXTURE_HPP_
+#define TEMPORARY_DIRECTORY_FIXTURE_HPP_
+
+#include <gmock/gmock.h>
+
+#include <string>
+
+#ifdef _WIN32
+# include <direct.h>
+# include <Windows.h>
+#else
+# include <unistd.h>
+# include <sys/types.h>
+# include <dirent.h>
+#endif
+
+
+class TemporaryDirectoryFixture : public ::testing::Test
+{
+public:
+  TemporaryDirectoryFixture()
+  {
+    char template_char[] = "tmp_test_dir.XXXXXX";
```

Comment:
:warning: This code will segfault.

This literal cannot be modified, you should not put it in a const and pass that to function expecting a `char *` as they will modfiy it.

See the longer version here: https://stackoverflow.com/a/31846777/1442517

Reply:
Interesting, would the better approach be to use a  `std::vector<char>` or a `std::unique_ptr<char[]>` instead?

Reply:
I think the correct solution is to strcpy "tmp_test_dir.XXXXXX" into a `char[]` of appropriate size.

So something like:
```c++
const template_char[SOME_MAX_SIZE];  // SOME_MAX_SIZE should be larger than the template string.
strcpy(template_char, "tmp_test_dir.XXXXXX");
```

Reply:
I am wrong. @zmichaels11 no need to specify the string length, this is brittle:
According to https://www.geeksforgeeks.org/char-vs-stdstring-vs-char-c/ and my tests:

```c++
char * foo = "abcd";
foo[0] = 'e'; // segv

char foo2[] = "abcd";
foo2[0] = 'e'; // no segv
```

so it seems safe. @zmichaels11 does that match what you saw?

See `-Wwrite-strings` GCC flag.

Reply:
We could repro with @zmichaels11 - the initial code works as attended and is correct

Reply:
Link to reproduction: https://github.com/ros2/rosbag2/pull/239#issuecomment-567690147